### PR TITLE
feat: refactor new error types (hs, indicative, itercom, iterable, drip)

### PR DESCRIPTION
**Pull request by ujjwal-ab**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/drip/util.js [hunk 1]
```diff
@@ -33,7 +35,7 @@ const userExists = async (Config, id) => {
     if (response && response.status) {
       return response.status === 200;
     }
-    throw new CustomError("Invalid response.");
+    throw new NetworkInstrumentationError("Invalid response.");
```
Comment (saikumarrs) [inline src/v0/destinations/drip/util.js:38]: Is this clearly an instrumentation error?

File: src/v0/destinations/hs/HSTransform-v1.js [hunk 2]
```diff
@@ -232,7 +234,7 @@ const batchIdentifyForrETL = (
         metadata.push(ev.metadata);
       });
     } else {
-      throw new CustomError("[HS]:: rETL -  Unknow hubspot operation", 400);
+      throw new TransformationError("rETL -  Unknow hubspot operation", 400);
```
Comment (saikumarrs) [inline src/v0/destinations/hs/HSTransform-v1.js:237]: This should not be a transformation error. Check the logic once again.

File: src/v0/destinations/hs/HSTransform-v2.js [hunk 3]
```diff
@@ -109,7 +113,7 @@ const processIdentify = async (message, destination, propertyMap) => {
   ) {
     addExternalIdToTraits(message);
     if (!objectType) {
-      throw new CustomError("objectType not found", 400);
+      throw new TransformationError("objectType not found");
```
Comment (saikumarrs) [inline src/v0/destinations/hs/HSTransform-v2.js:116]: Not a transformation error.

File: src/v0/destinations/hs/util.js [hunk 4]
```diff
@@ -476,9 +492,8 @@ const getExistingData = async (inputs, destination) => {
     searchResponse = processAxiosResponse(searchResponse);
 
     if (searchResponse.status !== 200) {
-      throw new CustomError(
-        `[HS]:: rETL - Error during searching object record. ${searchResponse.response?.message}`,
-        searchResponse.status || 400
+      throw new NetworkInstrumentationError(
```
Comment (saikumarrs) [inline src/v0/destinations/hs/util.js:495]: Incorrect error type.

File: src/v0/destinations/intercom/transform.js [hunk 5]
```diff
@@ -166,10 +173,7 @@ function process(event) {
   try {
     response = processSingleMessage(event.message, event.destination);
   } catch (error) {
-    throw new CustomError(
-      error.message || "Unknown error",
-      error.status || 400
-    );
+    throw new TransformationError(error.message || "Unknown error");
```
Comment (saikumarrs) [inline src/v0/destinations/intercom/transform.js:176]: Remove error handling here.

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: src/v0/destinations/drip/util.js (hunk 1 above)
Comment (ujjwal-ab) [reply src/v0/destinations/drip/util.js:38]: Yeah, it should be as it is the case when the response status is not 200.

Comment (saikumarrs) [reply src/v0/destinations/drip/util.js:38]: That's not the criteria for the network instrumentation error.

[the remark this answers is not in the document]
Comment (utsabc) [reply src/v0/destinations/drip/util.js:38]: lets change to Network error and pass the details (my proposal)

File: src/v0/destinations/hs/util.js (hunk 4 above)
Comment (ujjwal-ab) [reply src/v0/destinations/hs/util.js:495]: Should we throw NetworkError here?

File: src/v0/destinations/hs/HSTransform-v1.js (hunk 2 above)
Comment (ujjwal-ab) [reply src/v0/destinations/hs/HSTransform-v1.js:237]: We are setting that field by default. What should be the correct error category then?

**Review by utsabc — APPROVED** [review]

File: src/v0/destinations/hs/HSTransform-v2.js [hunk 6]
```diff
@@ -430,7 +430,7 @@ const batchEvents = destEvents => {
           associationObjectsEventChunk.push(event);
         }
       } else {
-        throw new InstrumentationError("rETL -  Error in getting operation");
+        throw new TransformationError("rETL -  Error in getting operation");
```
Comment (saikumarrs) [inline src/v0/destinations/hs/HSTransform-v2.js:433]: This should be an instrumentation error. The operation field is fetched from the message.

Comment (ujjwal-ab) [reply src/v0/destinations/hs/HSTransform-v2.js:433]: Yeah, It should be.

Comment (utsabc) [reply src/v0/destinations/hs/HSTransform-v2.js:433]: This `operation` field is set by us only in the previous processing flow if this is missing its not a fault of customer its a code bug

File: src/v0/destinations/hs/HSTransform-v2.js [hunk 7]
```diff
@@ -428,7 +430,7 @@ const batchEvents = destEvents => {
           associationObjectsEventChunk.push(event);
         }
       } else {
-        throw new CustomError("[HS]:: rETL -  Error in getting operation", 400);
+        throw new InstrumentationError("rETL -  Error in getting operation");
```
[the remark this answers is not in the document]
Comment (utsabc) [reply src/v0/destinations/hs/HSTransform-v2.js]: Same this is set in the code this is not a message `property`

[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/v0/destinations/hs/HSTransform-v2.js]: Actually, both the comments are on same [line](https://github.com/rudderlabs/rudder-transformer/pull/1672#discussion_r1048404510).

[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply src/v0/destinations/hs/HSTransform-v2.js]: [Here](https://github.com/rudderlabs/rudder-transformer/blob/ec99103a0620d7e43946ca9b14a71a5531c2147f/src/v0/destinations/hs/HSTransform-v2.js#L394), we are taking it from the message.

File: src/v0/destinations/hs/HSTransform-v2.js (hunk 6 above)
Comment (saikumarrs) [reply src/v0/destinations/hs/HSTransform-v2.js:433]: Oops. That's messed up logic. If @utsabc is sure, let's go back to `TransformationError` only.

Comment (ujjwal-ab) [reply src/v0/destinations/hs/HSTransform-v2.js:433]: But, what about [this](https://github.com/rudderlabs/rudder-transformer/blob/ec99103a0620d7e43946ca9b14a71a5531c2147f/src/v0/destinations/hs/HSTransform-v2.js#L394)?

Comment (saikumarrs) [reply src/v0/destinations/hs/HSTransform-v2.js:433]: That's what I was referring to earlier but @utsabc says we are setting that ourselves in the code somewhere.

Comment (ujjwal-ab) [reply src/v0/destinations/hs/HSTransform-v2.js:433]: Both are for different functions.

File: src/v0/destinations/hs/HSTransform-v2.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (utsabc) [reply src/v0/destinations/hs/HSTransform-v2.js]: Ujjwal but those are transformed event

[the remark this answers is not in the document]
Comment (utsabc) [reply src/v0/destinations/hs/HSTransform-v2.js]: We have already set those in processor flow

**Review by saikumarrs — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Pull request merged**